napi/parser/src/types.rs

```diff
@@ -22,13 +25,38 @@ pub struct ParserOptions {
     pub preserve_parens: Option<bool>,
 }
 
-#[napi(object)]
+#[napi]
 pub struct ParseResult {
-    #[napi(ts_type = "import(\"@oxc-project/types\").Program")]
-    pub program: String,
-    pub module: EcmaScriptModule,
-    pub comments: Vec<Comment>,
-    pub errors: Vec<Error>,
+    pub(crate) source_text: String,
+    // #[napi(ts_type = "import(\"@oxc-project/types\").Program")]
+    pub(crate) program: String,
+    pub(crate) module: EcmaScriptModule,
+    pub(crate) comments: Vec<Comment>,
+    pub(crate) errors: Vec<Error>,
+}
+
+#[napi]
+impl ParseResult {
+    #[napi(ts_return_type = "import(\"@oxc-project/types\").Program")]
+    pub fn ast(&mut self) -> String {
+        mem::take(&mut self.program)
+    }
+
+    pub fn module(&mut self) -> EcmaScriptModule {
+        mem::take(&mut self.module)
+    }
+
+    pub fn comments(&mut self) -> Vec<Comment> {
+        mem::take(&mut self.comments)
+    }
+
+    pub fn errors(&mut self) -> Vec<Error> {
+        mem::take(&mut self.errors)
+    }
+
+    pub fn magic_string(&mut self) -> MagicString {
+        MagicString::new(mem::take(&mut self.source_text))
+    }
```

Comment:
@Brooooooklyn is there any better approach to return this magic string instance without changing `ParseResult` from a `#[napi(object)]` to `#[napi]` class?

Reply:
Perhaps changing this to a class is better, I can add methods for getting line column numbers, and other string utility methods.

I'll just need to change this to `ParserReturn` with a result object.

Reply:
We can add `#[napi(getter)]` to the `ParserReturn` fields rather than the `methods`:

```rust
#[napi]
impl ParseResult {
    #[napi(getter, ts_return_type = "import(\"@oxc-project/types\").Program")]
    pub fn get_ast(&mut self) -> String {
        mem::take(&mut self.program)
    }
}
```

Generated .d.ts should be:

```ts
export class PasreResult {
  get ast(): import("@oxc-project/types").Program
}
```